test/e2e/federation-authn.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = framework.KubeDescribe("[Feature:Federation]", func() {
+	f := framework.NewDefaultFederatedFramework("federation-apiserver-authn")
+
+	var _ = Describe("Federation API server authentication", func() {
+		BeforeEach(func() {
+			framework.SkipUnlessFederated(f.Client)
+		})
+
+		It("should accept cluster resources when the client has right authentication credentials", func() {
+			framework.SkipUnlessFederated(f.Client)
+
+			svc := createServiceOrFail(f.FederationClientset_1_4, f.Namespace.Name)
+			deleteServiceOrFail(f.FederationClientset_1_4, f.Namespace.Name, svc.Name)
+		})
+
+		It("should not accept cluster resources when the client has invalid authentication credentials", func() {
+			framework.SkipUnlessFederated(f.Client)
+
+			contexts := f.GetUnderlyingFederatedContexts()
+
+			fcs, err := invalidAuthFederationClientSet(contexts[0].User)
```

Comment:
context[0] cannot be a federation API server. If it is, our tests don't work. `contexts` is obtained from calling `f.GetUnderlyingFederatedContexts()` which in turn checks that the contexts it returns are not federation API server's context/.